### ISSUE TITLE
"ember serve" fails after "ember build"

### ISSUE THREAD

**Issue by jgoclawski** [ember-cli/ember-cli#5480]
We have a Docker-based ember project that used to work with `ember-cli@2.3.0-beta.1` and before and fails with `ember-cli@2.3.0-beta.2` and `ember-cli@2.3.0`.
In a nutshell we have a Docker image that builds production version of Ember app, but we also use it for development. So let'a sssume we use it like that:
```  ember build --environment="production" ```
```  ember serve ``` (when running the image for development, mounting app volumes, etc.)

Usually the error below pops out, rarely the setup works like before, so it's not deterministic.

What solves the problem for us is deleting the "dist/" directory before running "ember serve". 
Another thing that solves the problem is running "ember serve" again in the container. So it seems that after the first build something gets left behind in the filesystem, which fixes the problem on second run.

Is it by design or is it a bug introduced between beta.1 and beta.2?

Output from shell:
```
root@58858076ddc2:/app# rm -r dist/
root@58858076ddc2:/app# ember build --environment="production"
version: 2.3.0
Built project successfully. Stored in "dist/".
root@58858076ddc2:/app# ember serve
version: 2.3.0
Livereload server on http://localhost:49152
Serving on http://localhost:4200/beta/
EEXIST: file already exists, mkdir '/app/dist/assets'
Error: EEXIST: file already exists, mkdir '/app/dist/assets'
    at Error (native)
    at Object.fs.mkdirSync (fs.js:794:18)
    at TreeSync.<anonymous> (/app/node_modules/ember-cli/node_modules/tree-sync/index.js:63:19)
    at Array.forEach (native)
    at TreeSync.sync (/app/node_modules/ember-cli/node_modules/tree-sync/index.js:50:14)
    at Class.module.exports.Task.extend.copyToOutputPath (/app/node_modules/ember-cli/lib/models/builder.js:124:10)
    at /app/node_modules/ember-cli/lib/models/builder.js:132:21
    at lib$rsvp$$internal$$tryCatch (/app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1036:16)
    at lib$rsvp$$internal$$invokeCallback (/app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1048:17)
    at /app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:331:11
^Croot@58858076ddc2:/app# ember serve
version: 2.3.0
Livereload server on http://localhost:49152
Serving on http://localhost:4200/beta/

Build successful - 1287ms.

Slowest Trees                                 | Total               
----------------------------------------------+---------------------
SourceMapConcat: Concat: Vendor               | 99ms                
LessCompiler                                  | 92ms                
Babel                                         | 65ms                

Slowest Trees (cumulative)                    | Total (avg)         
----------------------------------------------+---------------------
Babel (13)                                    | 296ms (22 ms)       
SourceMapConcat: Concat: Vendor (1)           | 99ms                
LessCompiler (1)                              | 92ms                
TemplateCompiler (2)                          | 86ms (43 ms)
```

**Comment by rwjblue** [comment]
We introduced a new mechanism to sync output trees in 2.3.0-beta.2 that avoids copying all files into dist on every rebuild, and instead only performs new operations. 

This error is coming from the mkdir operation here: https://github.com/stefanpenner/tree-sync/blob/master/index.js#L63

Questions:

* Before the first build (the one that always errors) is started, does dist/assets folder exist? Does it have files in it?
* I released an update yesterday for tree-sync, can you confirm which version you are using?

**Comment by stefanpenner** [comment]
@rwjblue it is safe for tree-sync mkdirSync to be try/catch and assert that the catch raised only a EEXIST, that operation (ensuring a DIR exists) is totally reasonable to allow in racy conditions. As is unlinkSync and rmdir, what isn't ok is change but that one will be inherently last-write-wins.

**Comment by rwjblue** [comment]
Thanks @stefanpenner. I'll work up a failing test + fix PR in tree-sync.

**Comment by rwjblue** [comment]
I submitted a test (that passes without changes) that shows an output directory that contains an existing directory works properly: https://github.com/stefanpenner/tree-sync/pull/5.

We will definitely need more information or a reproduction to be able to track this down...

**Comment by stefanpenner** [comment]
@rwjblue my guess is both happening concurrently, may allow for strange intermediate states.

**Comment by rwjblue** [comment]
@stefanpenner - Yes, that makes sense (and explains the error a bit). The repro steps were for `ember b` then `ember s` (in sequence not parallel), so I'm not sure...

**Comment by stefanpenner** [comment]
ya. strange. I think my above example is for a likely unrelated failure scenario

**Comment by jgoclawski** [comment]
Thanks for looking into it!
The commands were run sequentially, as in the example (unless there was a background process spawned). 

> Questions:
> - Before the first build (the one that always errors) is started, does dist/assets folder exist? Does it have files in it?

The first step in the example is `rm -r dist/`, to ensure it's empty.
I'm enclosing more output, to show that what fixes the problem apparently is that first `ember serve` deletes the `assets` subdirectory:

```
root@cc8e59029a63:/app# rm -r dist/
root@cc8e59029a63:/app# ls
Dockerfile  app  bower.json  bower_components  config  ember-cli-build.js  node_modules  package.json  public  testem.json  tests  tmp	vendor
root@cc8e59029a63:/app# ember build --environment="production"
version: 2.3.0
Built project successfully. Stored in "dist/".
root@cc8e59029a63:/app# ls
Dockerfile  app  bower.json  bower_components  config  dist  ember-cli-build.js  node_modules  package.json  public  testem.json  tests  tmp  vendor
root@cc8e59029a63:/app# ls dist/
assets	crossdomain.xml  fonts	form_icons  img  index.html  robots.txt
root@cc8e59029a63:/app# ls dist/assets/
ember-app-b440987750589a5673241097a76bcadb.js  ember-app-f4770bac19c2b875fdb556f71636f447.css  vendor-44e0dee0d3d6d08ad82ade552bd31c94.js  vendor-467b695a900f15aa450ecc6af98820f5.css
root@cc8e59029a63:/app# ember serve
version: 2.3.0
Livereload server on http://localhost:49152
Serving on http://localhost:4200/beta/
EEXIST: file already exists, mkdir '/app/dist/assets'
Error: EEXIST: file already exists, mkdir '/app/dist/assets'
    at Error (native)
    at Object.fs.mkdirSync (fs.js:794:18)
    at TreeSync.<anonymous> (/app/node_modules/ember-cli/node_modules/tree-sync/index.js:63:19)
    at Array.forEach (native)
    at TreeSync.sync (/app/node_modules/ember-cli/node_modules/tree-sync/index.js:50:14)
    at Class.module.exports.Task.extend.copyToOutputPath (/app/node_modules/ember-cli/lib/models/builder.js:124:10)
    at /app/node_modules/ember-cli/lib/models/builder.js:132:21
    at lib$rsvp$$internal$$tryCatch (/app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1036:16)
    at lib$rsvp$$internal$$invokeCallback (/app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1048:17)
    at /app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:331:11
^Croot@cc8e59029a63:/app# ls /app/dist/   
crossdomain.xml  fonts	index.html  robots.txt
root@cc8e59029a63:/app# ember serve
version: 2.3.0
Livereload server on http://localhost:49152
Serving on http://localhost:4200/beta/

Build successful - 1416ms.

Slowest Trees                                 | Total               
----------------------------------------------+---------------------
SourceMapConcat: Concat: Vendor               | 125ms               
LessCompiler                                  | 96ms                

Slowest Trees (cumulative)                    | Total (avg)         
----------------------------------------------+---------------------
Babel (13)                                    | 323ms (24 ms)       
SourceMapConcat: Concat: Vendor (1)           | 125ms               
LessCompiler (1)                              | 96ms                
TemplateCompiler (2)                          | 86ms (43 ms)        

^Croot@cc8e59029a63:/app# ls dist/
assets	crossdomain.xml  fonts	form_icons  img  index.html  robots.txt  testem.js  tests
root@cc8e59029a63:/app# ls dist/assets/
ember-app.css	 ember-app.js   failed.png  test-loader.js    test-support.js   tests.js	 vendor.css  vendor.map
ember-app.css.map  ember-app.map  passed.png  test-support.css  test-support.map  tests.map  vendor.js
```

> - I released an update yesterday for tree-sync, can you confirm which version you are using?

```
root@cc8e59029a63:/app# npm list tree-sync
npm info it worked if it ends with ok
npm info using npm@2.14.7
npm info using node@v4.2.2
ember-app@0.0.0 /app
└─┬ ember-cli@2.3.0
  └── tree-sync@1.0.2 

npm info ok
```

I will work on a small reproduction example, maybe it will be more helpful than above outputs :)

**Comment by jgoclawski** [comment]
I have tested the bug on my little TODO Ember app and I was able to reproduce it.

I've created a branch which shows the bug when you do `docker-compose build todo`.

Branch: https://github.com/jgoclawski/ember-tutorial/tree/build_serve_error

Diff: https://github.com/jgoclawski/ember-tutorial/commit/30e864f577309e6913b9e69059abf135e662f726

If you uncomment `RUN ember build -prod` then it works. You can play around in the container with `docker-compose run --rm todo bash` or just use it without Docker.

**Comment by jpadilla** [comment]
Just started running into this after upgrading from latest `2.2.0` beta to `2.3.0`.

**Comment by sheriffderek** [comment]
With _Ember and CLI_ at 2.4.2, after I build for production - if I `ember serve`

```
gold/ : ember serve
version: 2.4.2
Livereload server on http://localhost:49152
Serving on http://localhost:4200/
ENOENT: no such file or directory, open '/Users/sheriff/Sites/CLIENTS/gold/dist/assets/failed.png'
Error: ENOENT: no such file or directory, open '/Users/sheriff/Sites/CLIENTS/gold/dist/assets/failed.png'
    at Error (native)
    at Object.fs.openSync (fs.js:584:18)
    at Object.fs.writeFileSync (fs.js:1224:33)
    at TreeSync.<anonymous> (/Users/sheriff/Sites/CLIENTS/gold/node_modules/ember-cli/node_modules/tree-sync/index.js:59:19)
    at Array.forEach (native)
    at TreeSync.sync (/Users/sheriff/Sites/CLIENTS/gold/node_modules/ember-cli/node_modules/tree-sync/index.js:50:14)
    at Class.module.exports.Task.extend.copyToOutputPath (/Users/sheriff/Sites/CLIENTS/gold/node_modules/ember-cli/lib/models/builder.js:124:10)
    at /Users/sheriff/Sites/CLIENTS/gold/node_modules/ember-cli/lib/models/builder.js:132:21
    at lib$rsvp$$internal$$tryCatch (/Users/sheriff/Sites/CLIENTS/gold/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1036:16)
    at lib$rsvp$$internal$$invokeCallback (/Users/sheriff/Sites/CLIENTS/gold/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1048:17)
fs.js:60
      throw err;  // Forgot a callback but don't know where? Use NODE_DEBUG=fs
      ^

Error: ENOENT: no such file or directory, rmdir '/Users/sheriff/Sites/CLIENTS/gold/dist/assets'
    at Error (native)
gold/ :
```

If I `ember build` before `ember serve` then things work as expected. I'm not sure why the `/dist` folder would matter for `ember serve` though - since I can serve without ever building.

**Comment by kdhunter** [comment]
This fails reliably for me on a completely vanilla (right after "ember new" with no changes) project from ember 2.5.

KDH-MBP15:ember kevin$ ember version
ember-cli: 2.5.0
node: 0.12.14
os: darwin x64

KDH-MBP15:ember kevin$ ember new serve-bug
installing app
  create .bowerrc
  create .editorconfig
  create .ember-cli
  create .jshintrc
  create .travis.yml
  create .watchmanconfig
  create README.md
  create app/app.js
  create app/components/.gitkeep
  create app/controllers/.gitkeep
  create app/helpers/.gitkeep
  create app/index.html
  create app/models/.gitkeep
  create app/resolver.js
  create app/router.js
  create app/routes/.gitkeep
  create app/styles/app.css
  create app/templates/application.hbs
  create app/templates/components/.gitkeep
  create bower.json
  create config/environment.js
  create ember-cli-build.js
  create .gitignore
  create package.json
  create public/crossdomain.xml
  create public/robots.txt
  create testem.js
  create tests/.jshintrc
  create tests/helpers/destroy-app.js
  create tests/helpers/module-for-acceptance.js
  create tests/helpers/resolver.js
  create tests/helpers/start-app.js
  create tests/index.html
  create tests/integration/.gitkeep
  create tests/test-helper.js
  create tests/unit/.gitkeep
  create vendor/.gitkeep
Successfully initialized git.
Installed packages for tooling via npm.
Installed browser packages via Bower.

KDH-MBP15:ember kevin$ cd serve-bug
KDH-MBP15:serve-bug kevin$ ember build --environment=production
Built project successfully. Stored in "dist/".
File sizes:
 - serve-bug-63a11e94bb7c79d4fe4301b86993a5f1.js: 4.78 KB (1.46 KB gzipped)
 - serve-bug-d41d8cd98f00b204e9800998ecf8427e.css: 0 B
 - vendor-49028e44823ec2135e0139d2eb1f9d7b.js: 689.6 KB (179.82 KB gzipped)
 - vendor-d41d8cd98f00b204e9800998ecf8427e.css: 0 B

KDH-MBP15:serve-bug kevin$ find dist
dist
dist/assets
dist/assets/serve-bug-63a11e94bb7c79d4fe4301b86993a5f1.js
dist/assets/serve-bug-d41d8cd98f00b204e9800998ecf8427e.css
dist/assets/vendor-49028e44823ec2135e0139d2eb1f9d7b.js
dist/assets/vendor-d41d8cd98f00b204e9800998ecf8427e.css
dist/crossdomain.xml
dist/index.html
dist/robots.txt

KDH-MBP15:serve-bug kevin$ ember serve
Livereload server on http://localhost:49154
Serving on http://localhost:4200/
ENOENT, no such file or directory '/private/tmp/ember/serve-bug/dist/assets/failed.png'
Error: ENOENT, no such file or directory '/private/tmp/ember/serve-bug/dist/assets/failed.png'
    at Error (native)
    at Object.fs.openSync (fs.js:500:18)
    at Object.fs.writeFileSync (fs.js:1099:15)
    at TreeSync.<anonymous> (/private/tmp/ember/serve-bug/node_modules/ember-cli/node_modules/tree-sync/index.js:59:19)
    at Array.forEach (native)
    at TreeSync.sync (/private/tmp/ember/serve-bug/node_modules/ember-cli/node_modules/tree-sync/index.js:50:14)
    at Class.module.exports.Task.extend.copyToOutputPath (/private/tmp/ember/serve-bug/node_modules/ember-cli/lib/models/builder.js:124:10)
    at /private/tmp/ember/serve-bug/node_modules/ember-cli/lib/models/builder.js:132:21
    at lib$rsvp$$internal$$tryCatch (/private/tmp/ember/serve-bug/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1036:16)
    at lib$rsvp$$internal$$invokeCallback (/private/tmp/ember/serve-bug/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1048:17)
fs.js:77
      throw err;  // Forgot a callback but don't know where? Use NODE_DEBUG=fs
            ^
Error: ENOENT, rmdir '/private/tmp/ember/serve-bug/dist/assets'
    at Error (native)


KDH-MBP15:serve-bug kevin$ find dist
dist
dist/crossdomain.xml
dist/index.html
dist/robots.txt
dist/testem.js
dist/tests


KDH-MBP15:serve-bug kevin$ ember serve
Livereload server on http://localhost:49154
Serving on http://localhost:4200/

Build successful - 2208ms.

Slowest Trees                                 | Total               
----------------------------------------------+---------------------
Babel                                         | 436ms               
Babel                                         | 332ms               
SourceMapConcat: Concat: Addon JS             | 123ms               

Slowest Trees (cumulative)                    | Total (avg)         
----------------------------------------------+---------------------
Babel (12)                                    | 1178ms (98 ms)      
SourceMapConcat: Concat: Addon JS (1)         | 123ms    


^CKDH-MBP15:serve-bug kevin$ find dist
dist
dist/assets
dist/assets/failed.png
dist/assets/passed.png
dist/assets/serve-bug.css
dist/assets/serve-bug.js
dist/assets/serve-bug.map
dist/assets/test-loader.js
dist/assets/test-support.css
dist/assets/test-support.js
dist/assets/test-support.map
dist/assets/tests.js
dist/assets/tests.map
dist/assets/vendor.css
dist/assets/vendor.js
dist/assets/vendor.map
dist/crossdomain.xml
dist/index.html
dist/robots.txt
dist/testem.js
dist/tests
dist/tests/index.html
           


The first call to `ember serve` dies, but in doing so it alters the contents of the `dist` folder sufficiently that the subsequent call to `ember serve` succeeds.  Note that the file that `ember serve` is looking for in my case is the `failed.png` that is normally there for test results.

**Comment by nathanhammond** [comment]
```sh
ember new serve-bug
cd serve-bug
ember build --environment=production
ember serve
```

This works for me in the latest Ember CLI beta. Closing as resolved.

**Comment by jgoclawski** [comment]
Thanks, it's resolved for me as well (tried on 2.7.0-beta.4).
